src/core/ComboBox/ComboBox.tsx

```diff
@@ -180,189 +174,274 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
-  );
-
-  // Maintain internal selected value state synced with `value` prop
-  const [selectedValue, setSelectedValue] = React.useState<T | undefined>(
-    value,
-  );
-  React.useEffect(() => {
-    setSelectedValue(value);
-  }, [value]);
-
-  // Controlled input value
-  const [inputValue, setInputValue] = React.useState<string>('');
-  const onInput = React.useCallback(
+  // Filter options based on input value
+  const [inputValue, setInputValue] = React.useState(inputProps?.value ?? '');
+  const handleOnInput = React.useCallback(
     (event: React.ChangeEvent<HTMLInputElement>) => {
-      setInputValue(event.target.value);
+      const { value } = event.currentTarget;
+      setInputValue(value);
+      dispatch(['open']);
+      setFilteredOptions(
+        filterFunction?.(options, value) ??
+          options.filter((option) =>
+            option.label.toLowerCase().includes(value.toLowerCase()),
+          ),
+      );
       inputProps?.onChange?.(event);
     },
-    [inputProps],
+    [filterFunction, inputProps, options],
   );
 
-  // update inputValue and focusedIndex every time selected value changes
+  // Reset focused item when filter changes
   React.useEffect(() => {
-    const selectedOption = options.find(({ value }) => value === selectedValue);
-    setInputValue(selectedOption?.label ?? '');
-    setFocusedIndex(selectedOption ? options.indexOf(selectedOption) : -1);
-  }, [selectedValue, options]);
+    dispatch(['focus']);
+  }, [filteredOptions]);
 
-  // Filter options and update focus when input value changes
+  // When the value prop changes, update the selectedIndex
   React.useEffect(() => {
-    if (!isOpen) {
-      return;
-    }
+    dispatch([
+      'select',
+      options.findIndex((option) => option.value === valueProp),
+    ]);
+  }, [options, valueProp]);
 
-    // if input is empty or same as selected value, show the whole list
-    const selectedOption = options.find(({ value }) => value === selectedValue);
-    if (!inputValue || selectedOption?.label === inputValue) {
+  // When value is selected, update input value and reset filtered options
+  React.useEffect(() => {
+    if (selectedIndex != undefined && selectedIndex >= 0) {
+      setInputValue(options[selectedIndex]?.label ?? '');
       setFilteredOptions(options);
-      return;
     }
+  }, [options, selectedIndex]);
+
+  // When dropdown is closed, reset input value to selected option label
+  React.useEffect(() => {
+    if (!isOpen && selectedIndex != undefined && selectedIndex >= 0) {
+      setInputValue(options[selectedIndex]?.label ?? '');
+    }
+  }, [isOpen, options, selectedIndex]);
+
+  // Call user-defined onChange when the value actually changes
+  React.useEffect(() => {
+    if (selectedIndex != undefined && selectedIndex >= 0) {
+      const value = options[selectedIndex]?.value;
+      if (value === valueProp) {
+        return;
+      }
+      onChangeProp.current?.(value);
```

Comment:
It will be also called on options change and on selection, not only on value change.